lib/Lower/ConvertExpr.cpp

```diff
@@ -795,66 +798,89 @@ class ExprLowering {
+    // Lower arguments
+    // For now, logical arguments for intrinsic are lowered to `fir.logical`
+    // so that TRANSFER can work. For some arguments, it could lead to useless
+    // conversions (e.g scalar MASK of MERGE will be converted to `i1`), but
+    // the generated code is at least correct. To improve this, the intrinsic
+    // lowering facility should control argument lowering.
+    auto restorer = Fortran::common::ScopedSet(genLogicalAsI1, false);
+    for (const auto &arg : procRef.arguments()) {
+      if (auto *expr = Fortran::evaluate::UnwrapExpr<
+              Fortran::evaluate::Expr<Fortran::evaluate::SomeType>>(arg)) {
+        operands.push_back(genval(*expr));
+      } else {
+        operands.push_back(nullptr); // optional
       }
-      // Let the intrinsic library lower the intrinsic function call
-      llvm::StringRef name{intrinsic->name};
-      return intrinsics.genval(getLoc(), builder, name, ty, operands);
-    } else {
-      // implicit interface implementation only
-      // TODO: explicit interface
-      llvm::SmallVector<mlir::Type, 2> argTypes;
-      llvm::SmallVector<mlir::Value, 2> operands;
-      // Logical arguments of user functions must be lowered to `fir.logical`
-      // and not `i1`.
-      auto restorer = Fortran::common::ScopedSet(genLogicalAsI1, false);
-      for (const auto &arg : funRef.arguments()) {
-        assert(arg.has_value() &&
-               "optional argument requires explicit interface");
-        const auto *expr = arg->UnwrapExpr();
-        assert(expr && "assumed type argument requires explicit interface");
-        if (const auto *sym =
-                Fortran::evaluate::UnwrapWholeSymbolDataRef(*expr)) {
-          mlir::Value argRef = symMap.lookupSymbol(*sym);
-          assert(argRef && "could not get symbol reference");
-          argTypes.push_back(argRef.getType());
-          operands.push_back(argRef);
-        } else {
-          // TODO create temps for expressions
-          TODO();
-        }
+    }
+    // Let the intrinsic library lower the intrinsic procedure call
+    llvm::StringRef name{intrinsic.name};
+    return intrinsics.genval(getLoc(), builder, name, resultType[0], operands);
+  }
+
+  mlir::Value genProcedureRef(const Fortran::evaluate::ProcedureRef procRef,
+                              mlir::ArrayRef<mlir::Type> resultType) {
+    if (const auto *intrinsic{procRef.proc().GetSpecificIntrinsic()}) {
+      return genIntrinsicRef(procRef, *intrinsic, resultType[0]);
+    }
+    // Implicit interface implementation only
+    // TODO: Explicit interface, we need to use Characterize here,
+    // evaluate::IntrinsicProcTable is required to use it.
+    llvm::SmallVector<mlir::Type, 2> argTypes;
+    llvm::SmallVector<mlir::Value, 2> operands;
+    // Logical arguments of user functions must be lowered to `fir.logical`
+    // and not `i1`.
+    auto restorer = Fortran::common::ScopedSet(genLogicalAsI1, false);
```

Comment:
Maybe `/*parameterName=*/false` to clarify?

Reply:
you mean, to indicate that this is setting `genLogicalAsI1` to false ?

Reply:
I find this interface confusing without reading the ScopedSet code.